libcontainer/cgroups/fs/cpu.go

```diff
@@ -71,15 +73,32 @@ func (s *CpuGroup) Set(path string, r *configs.Resources) error {
 			return fmt.Errorf("the minimum allowed cpu-shares is %d", sharesRead)
 		}
 	}
+
+	var period string
 	if r.CpuPeriod != 0 {
-		if err := cgroups.WriteFile(path, "cpu.cfs_period_us", strconv.FormatUint(r.CpuPeriod, 10)); err != nil {
-			return err
+		period = strconv.FormatUint(r.CpuPeriod, 10)
+		if err := cgroups.WriteFile(path, "cpu.cfs_period_us", period); err != nil {
+			// Sometimes when the period to be set is smaller
+			// than the current one, it is rejected by the kernel
+			// (EINVAL) as old_quota/new_period exceeds the parent
+			// cgroup quota limit. If this happens, ignore the
+			// error for now and retry after setting the quota.
+			if !errors.Is(err, unix.EINVAL) {
+				return err
+			}
```

Comment:
If Quota is not being set (r.CpuQuota == 0), this will silently fail Period change as well. Any idea whether it's possible to get EINVAL in cases other than the original bug report?

Reply:
Good catch. Funny that I added quite a few corner cases for cpu quota (e.g. in #2458) and it's still not enough :)

Reply:
Fixed, and added yet another corner test case ("setting invalid period without quota should fail").